staging/src/k8s.io/apiserver/pkg/util/peerproxy/peer_discovery.go

```diff
@@ -0,0 +1,227 @@
+
+func (h *peerProxyHandler) updatePeerDiscoveryInfo(oldObj interface{}, newObj interface{}) {
+	oldLease, ok := oldObj.(*v1.Lease)
+	if !ok {
+		klog.Errorf("invalid lease object provided to updatePeerServedResources(), received type: %T", oldObj)
+		return
+	}
+
+	newLease, ok := newObj.(*v1.Lease)
+	if !ok {
+		klog.Errorf("invalid lease object provided to updatePeerServedResources(), received type: %T", newObj)
+		return
+	}
+
+	switch {
+	case newLease == nil && oldLease == nil:
+		return
+	case newLease != nil && oldLease != nil:
+		if oldLease.Spec.HolderIdentity == nil {
+			klog.Errorf("invalid lease object provided to updatePeerServedResources(), missing holderIdentity, received obj: %v", oldLease)
+			return
+		}
+		if newLease.Spec.HolderIdentity == nil {
+			klog.Errorf("invalid lease object provided to updatePeerServedResources(), missing holderIdentity, received obj: %v", newLease)
+			return
+		}
+		// Delete old discovery info if holderIdentity changed, implying
+		// the sever was restarted.
+		if oldLease.Name == newLease.Name && *oldLease.Spec.HolderIdentity != *newLease.Spec.HolderIdentity {
+			h.withPeerAggDiscoveryCacheLock(func(cache map[string]*peerAggDiscoveryInfo) {
+				delete(cache, oldLease.Name)
+			})
+		}
+
+		if err := h.addPeerDiscoveryInfoToCache(newLease); err != nil {
+			klog.ErrorS(err, "error adding peer to discovery cache", "serverID", newLease.Name)
+		}
+	case oldLease == nil && newLease != nil:
+		if err := h.addPeerDiscoveryInfoToCache(newLease); err != nil {
+			klog.ErrorS(err, "error adding peer to discovery cache", "serverID", newLease.Name)
+		}
+	case oldLease != nil && newLease == nil:
+		h.withPeerAggDiscoveryCacheLock(func(cache map[string]*peerAggDiscoveryInfo) {
+			delete(cache, oldLease.Name)
+		})
+	}
+}
+
+func (h *peerProxyHandler) addPeerDiscoveryInfoToCache(serverIdentityLease *v1.Lease) error {
+	if serverIdentityLease == nil || serverIdentityLease.Spec.HolderIdentity == nil {
+		return fmt.Errorf("error adding peer discovery to cache, received nil lease or missing holderIdentity, lease obj: %v", serverIdentityLease)
+	}
+
+	// Don't record our own discovery info.
+	if serverIdentityLease.Name == h.serverId {
+		return nil
+	}
+
+	var discoveryDoc *peerAggDiscoveryInfo
+	var ok bool
+	h.withPeerAggDiscoveryCacheRLock(func(cache map[string]*peerAggDiscoveryInfo) {
+		discoveryDoc, ok = cache[serverIdentityLease.Name]
+	})
+	if !ok {
+		return h.fetchNewDiscoveryFor(serverIdentityLease.Name, serverIdentityLease.Spec.HolderIdentity)
+	}
+
+	if len(discoveryDoc.servedResources) == 0 {
+		return h.fetchNewDiscoveryFor(serverIdentityLease.Name, serverIdentityLease.Spec.HolderIdentity)
+	}
+
+	if discoveryDoc.holderIdentity != *serverIdentityLease.Spec.HolderIdentity {
+		return h.fetchNewDiscoveryFor(serverIdentityLease.Name, serverIdentityLease.Spec.HolderIdentity)
+	}
+
+	return nil
+}
+
+func (h *peerProxyHandler) fetchNewDiscoveryFor(serverID string, holderIdentity *string) error {
```

Comment:
nit: Can we take a string instead of pointer here for holderIdentity? Would make it more clear that it cannot be nil.

Reply:
Done